File: react-native/library/src/starPlanet/components/TspStarRating.js
import React from 'react';
import { Pressable, Text, View, StyleSheet } from 'react-native';
import { styles as shared, withTheme } from '../utils/shared';

/** Interactive star rating. */
export function TspStarRating({
  value = 0,
  max = 5,
  disabled = false,
  variant = 'default',
  theme,
  onChange,
}) {
  const t = withTheme(theme);
  const count = Math.max(1, Math.min(10, Number(max) || 5));
  const readonly = variant === 'readonly' || disabled;
  const stars = [];
  for (let i = 1; i <= count; i += 1) {
    const filled = i <= value;
    stars.push(
      <Pressable
        key={i}
        accessibilityRole="button"
        accessibilityLabel={`${i}`}
        disabled={readonly}
        onPress={() => onChange?.(i)}
        style={local.starHit}
      >
        <Text style={[local.star, { color: filled ? t.brandPrimary : t.borderDefault }]}>★</Text>
      </Pressable>
    );
  }
  return (
    <View
      accessibilityRole="adjustable"
      accessibilityValue={{ min: 0, max: count, now: value }}
      style={[local.root, disabled && shared.disabled]}
    >
      {stars}
    </View>
  );
}

const local = StyleSheet.create({
  root: { flexDirection: 'row', alignItems: 'center', gap: 4 },
  starHit: { padding: 2 },
  star: { fontSize: 22, lineHeight: 24 },
});

File: ios-swiftui/library/Sources/PlanetComponents/TspStarRating.swift
import SwiftUI

/// Interactive star rating.
public struct TspStarRating: View {
    public var value: Int
    public var max: Int
    public var disabled: Bool
    public var variant: String
    public var theme: StarPlanetTheme
    public var onChange: (Int) -> Void

    public init(
        value: Int = 0,
        max: Int = 5,
        disabled: Bool = false,
        variant: String = "default",
        theme: StarPlanetTheme = .sky,
        onChange: @escaping (Int) -> Void = { _ in }
    ) {
        self.value = value
        self.max = max
        self.disabled = disabled
        self.variant = variant
        self.theme = theme
        self.onChange = onChange
    }

    public var body: some View {
        let count = Swift.min(10, Swift.max(1, max))
        let readonly = variant == "readonly" || disabled
        HStack(spacing: 4) {
            ForEach(1...count, id: \.self) { i in
                Button {
                    guard !readonly else { return }
                    onChange(i)
                } label: {
                    Text("★")
                        .font(.system(size: 22))
                        .foregroundColor(i <= value ? theme.brandPrimary : theme.borderDefault)
                        .padding(2)
                }
                .buttonStyle(.plain)
                .disabled(readonly)
                .accessibilityLabel("\(i)")
            }
        }
        .opacity(disabled ? 0.45 : 1)
        .accessibilityValue(Text("\(value)"))
    }
}
